src/field_theory/perfect_closure.lean

```diff
@@ -18,65 +18,82 @@ open function
 
 section defs
 
-variables (K : Type u) [field K] (p : ℕ) [fact p.prime] [char_p K p]
+variables (R : Type u) [comm_semiring R] (p : ℕ) [fact p.prime] [char_p R p]
 
-/-- A perfect field is a field of characteristic p that has p-th root. -/
-class perfect_field (K : Type u) [field K] (p : ℕ) [fact p.prime] [char_p K p] : Type u :=
-(pth_root' : K → K)
-(frobenius_pth_root' : ∀ x, frobenius K p (pth_root' x) = x)
+/-- A perfect ring is a ring of characteristic p that has p-th root. -/
+class perfect_ring : Type u :=
+(pth_root' : R → R)
+(frobenius_pth_root' : ∀ x, frobenius R p (pth_root' x) = x)
+(pth_root_frobenius' : ∀ x, pth_root' (frobenius R p x) = x)
```

Comment:
Better or worse as follows?
```suggestion
(frobenius_pth_root' : function.left_inverse (frobenius R p) pth_root')
(pth_root_frobenius' : function.right_inverse (frobenius R p) pth_root')
```

Reply:
I would prefer the original one because it is more concrete.